flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/TableEnvironmentTest.scala

```diff
@@ -408,6 +408,330 @@ class TableEnvironmentTest {
+        |) WITH (
+        |  'connector' = 'COLLECTION',
+        |  'is-bounded' = 'false'
+        |)
+      """.stripMargin
+    tableEnv.executeSql(statement)
+
+    assertThatThrownBy(() => tableEnv.executeSql("""
+                                                   |ALTER TABLE MyTable MODIFY (
+                                                   |  x STRING FIRST
+                                                   |)
+                                                   |""".stripMargin))
+      .hasMessageContaining("Try to modify a column 'x' which does not exist in the table.")
+      .isInstanceOf(classOf[ValidationException])
+
+    assertThatThrownBy(() => tableEnv.executeSql("""
+                                                   |ALTER TABLE MyTable MODIFY (
+                                                   |  b INT FIRST,
+                                                   |  a BIGINT AFTER x
+                                                   |)
+                                                   |""".stripMargin))
+      .hasMessageContaining("Referenced column 'x' by 'AFTER' does not exist in the table.")
+      .isInstanceOf(classOf[ValidationException])
+
+    tableEnv.executeSql("""
+                          |ALTER TABLE MyTable MODIFY (
+                          |  e AS UPPER(c) FIRST,
+                          |  WATERMARK FOR ts1 AS ts1,
+                          |  a DOUBLE NOT NULL AFTER ts2,
+                          |  PRIMARY KEY(d) NOT ENFORCED
+                          |)
+                          |""".stripMargin)
+
+    val expectedResult = util.Arrays.asList(
+      Row.of("e", "STRING", Boolean.box(true), null, "AS UPPER(`c`)", null),
+      Row.of("b", "INT", Boolean.box(false), null, null, null),
+      Row.of("c", "STRING", Boolean.box(true), null, "METADATA VIRTUAL", null),
+      Row.of("d", "STRING", Boolean.box(false), "PRI(d)", null, null),
+      Row.of("ts1", "TIMESTAMP(3) *ROWTIME*", Boolean.box(true), null, null, "`ts1`"),
+      Row.of("ts2", "TIMESTAMP(3)", Boolean.box(true), null, null, null),
+      Row.of("a", "DOUBLE", Boolean.box(false), null, null, null)
+    )
+    val tableResult = tableEnv.executeSql("DESCRIBE MyTable")
+    assertEquals(ResultKind.SUCCESS_WITH_CONTENT, tableResult.getResultKind)
+    checkData(expectedResult.iterator(), tableResult.collect())
+  }
+
+  @Test
+  def testAlterTableModifyConstraint(): Unit = {
+    val statement =
+      """
+        |CREATE TABLE MyTable (
+        |  a BIGINT,
+        |  b INT,
+        |  c STRING METADATA VIRTUAL
+        |) WITH (
+        |  'connector' = 'COLLECTION',
+        |  'is-bounded' = 'false'
+        |)
+      """.stripMargin
+    tableEnv.executeSql(statement)
+
+    assertThatThrownBy(() => tableEnv.executeSql("""
+                                                   |ALTER TABLE MyTable MODIFY (
+                                                   |  PRIMARY KEY (x) NOT ENFORCED
```

Comment:
Does modifying primary key require the table also exist primary key, I think we can investigate MySQL behavior, and align with it. Watermark behavior is similar.